File: evaluators/exact.py
from evaluators.types import EvaluatorResult
from evaluators.utils import _run_evaluator

import json
from typing import Any


def exact_match(*, inputs: Any, outputs: Any, **kwargs: Any) -> EvaluatorResult:
    """
    Performs exact matching between input and output values.

    Args:
        inputs (Any): Inputs to compare
        outputs (Any): Outputs to compare

    Returns:
        MatchResult: Contains match result with score 1.0 for exact match, 0.0 otherwise
    """

    def get_score():
        if inputs is None or outputs is None:
            raise ValueError("Exact match requires both inputs and outputs")
        # Convert both to JSON strings for deep comparison
        inputs_json = json.dumps(inputs, sort_keys=True)
        outputs_json = json.dumps(outputs, sort_keys=True)
        return 1.0 if inputs_json == outputs_json else 0.0

    return _run_evaluator(
        run_name="exact_match", evaluator_fn=get_score, feedback_key="exact_match"
    )

File: js/src/evaluators/exact.ts
import { _runEvaluator } from "./utils.js";

const _scorer = (params: { inputs: unknown; outputs: unknown }) => {
  const { inputs, outputs } = params;
  if (inputs === null || outputs === null) {
    throw new Error("Exact match requires both inputs and outputs");
  }
  const inputsJson = JSON.stringify(inputs, null, 2);
  const outputsJson = JSON.stringify(outputs, null, 2);
  return inputsJson === outputsJson;
};

/**
 * Performs exact matching between input and output values.
 * @param inputs Inputs to compare
 * @param outputs Outputs to compare
 * @returns
 */
export const exactMatch = async (params: {
  inputs: unknown;
  outputs: unknown;
}) => {
  return _runEvaluator("exact_match", _scorer, "exact_match", params);
};
